## Data

In [15]:
import os
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
import sklearn
import pandas as pd
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RandomizedSearchCV
from sklearn.pipeline import FeatureUnion

In [16]:
titanic_path = os.path.join("datasets", "titanic")

def load_data(filename, path):
    csv_path = os.path.join(path, filename)
    return pd.read_csv(csv_path)

train_data = load_data("train.csv", titanic_path)
test_data = load_data("test.csv", titanic_path)

In [17]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [18]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


891 total entries. Columns 'age', 'cabin' and 'embarked' are not full i.e. sometimes null.
'cabin' has many null values so we will focus on 'age' and 'embarked' for now.

In [19]:
train_data['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [20]:
train_data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [21]:
train_data['Cabin'].value_counts()

B96 B98        4
G6             4
C23 C25 C27    4
C22 C26        3
F2             3
              ..
A10            1
A7             1
D10 D12        1
C106           1
C101           1
Name: Cabin, Length: 147, dtype: int64

In [22]:
train_data['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [23]:
train_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Feature Engineering

In [10]:
#Correlation
corr_matrix = train_data.corr()
corr_matrix["Survived"].sort_values(ascending=False)

Survived       1.000000
Fare           0.257307
Parch          0.081629
PassengerId   -0.005007
SibSp         -0.035322
Age           -0.077221
Pclass        -0.338481
Name: Survived, dtype: float64

In [70]:
def add_features(dataset, flag):
    #Family size feature
    dataset["FamilySize"] = dataset["SibSp"] + dataset["Parch"]

    #Binary feature whether or not the passanger was alone
    #dataset['Alone'] = dataset['FamilySize'].clip(upper=1)

    #Age range feature: 0-2, 3-5, 6-10, 11-18, 19-35, 36-65, 66-80.
    bins = [0, 2, 5, 10, 18, 40, 65, 80]
    names = [0, 1, 2, 3, 4, 5, 6, 7]
    d = dict(enumerate(names, 1))
    dataset['AgeRange'] = np.vectorize(d.get)(np.digitize(dataset['Age'], bins))

    #Title feature
    dataset['Title'] = dataset.Name.str.split(',', expand=True)[1].str.split('.',expand=True)[0]
    if flag == True:
        title_dict = {' Mr': 1, ' Mrs': 2, ' Dona':2, ' Miss':3, ' Master':4, ' Don':5, ' Rev':6, ' Dr':7, ' Mme':2,
                   ' Ms':9, ' Major':10, ' Lady':11, ' Sir':12, ' Mlle':3, ' Col':14, ' Capt':15,
                   ' the Countess':11, ' Jonkheer':11}
        dataset.replace({'Title': title_dict}, inplace=True)
    return dataset

#Copy dataset
train_data_copy = train_data.copy()
train_data_copy2 = train_data.copy()

train_data_copy = add_features(train_data_copy, False)
train_data_copy2 = add_features(train_data_copy2, True)

#Correlation
corr_matrix = train_data_copy.corr()
corr_matrix["Survived"].sort_values(ascending=False)

Survived       1.000000
Fare           0.257307
Parch          0.081629
FamilySize     0.016639
PassengerId   -0.005007
SibSp         -0.035322
Age           -0.077221
AgeRange      -0.152182
Pclass        -0.338481
Name: Survived, dtype: float64

In [76]:
train_data_copy.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  AgeRange  \
0      0         A/5 21171   7.2500   NaN        S           1         4   
1      0          PC 17599  71.2833   C85        C           1         4   
2      0  STON/O2. 3101282   7.9250   NaN        S           0         4   
3      0            113803  53.1000  C123        S           1         4   
4      0            373450   8.0500   NaN        S           0         4   

   Title  
0     Mr  
1    Mrs  
2   Miss  
3    Mrs  
4     Mr

In [77]:
train_data_copy2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  AgeRange  \
0      0         A/5 21171   7.2500   NaN        S           1         4   
1      0          PC 17599  71.2833   C85        C           1         4   
2      0  STON/O2. 3101282   7.9250   NaN        S           0         4   
3      0            113803  53.1000  C123        S           1         4   
4      0            373450   8.0500   NaN        S           0         4   

   Title  
0      1  
1      2  
2      3  
3      2  
4      1

## Preprocessing without new features

In [13]:
class DataSelector(BaseEstimator, TransformerMixin):
    def __init__(self, attribute_names):
        self.attribute_names = attribute_names
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        return X[self.attribute_names]
    
num_pipeline = Pipeline([
        ("select_numeric", DataSelector(["Age", "SibSp", "Parch", "Fare"])),
        ("imputer", SimpleImputer(strategy="median")),
    ])

#Imputer for string categories
class MostFrequentImputer(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.most_frequent_ = pd.Series([X[c].value_counts().index[0] for c in X],
                                        index=X.columns)
        return self
    def transform(self, X, y=None):
        return X.fillna(self.most_frequent_)

cat_pipeline = Pipeline([
        ("select_cat", DataSelector(["Pclass", "Sex", "Embarked"])),
        ("imputer", MostFrequentImputer()),
        ("cat_encoder", OneHotEncoder(sparse=False)),
    ])

preprocess_pipeline = FeatureUnion(transformer_list=[
        ("num_pipeline", num_pipeline),
        ("cat_pipeline", cat_pipeline),
    ])

X_train = preprocess_pipeline.fit_transform(train_data)
y_train = train_data["Survived"]

## Training without new features

In [14]:
forest_clf = RandomForestClassifier(n_estimators=100, random_state=42)
forest_clf.fit(X_train, y_train)
forest_scores = cross_val_score(forest_clf, X_train, y_train, cv=5)
forest_scores.mean()

0.8070154944327355

In [15]:
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 2000, num = 50)]
max_features = ['auto', 'sqrt', 'log2']
max_depth = [int(x) for x in np.linspace(10, 220, num = 11)]
max_depth.append(None)
min_samples_split = [2, 3, 4, 5, 6, 7, 8, 9, 10]
min_samples_leaf = [1, 2, 3, 4, 5, 6, 7]
bootstrap = [True, False]

random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

rnd_search = RandomizedSearchCV(forest_clf, param_distributions=random_grid,
                                n_iter=250, cv=5, scoring='neg_mean_squared_error',
                                verbose=2, random_state=42, n_jobs=-1)
rnd_search.fit(X_train, y_train)

Fitting 5 folds for each of 250 candidates, totalling 1250 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  17 tasks      | elapsed:    7.4s
[Parallel(n_jobs=-1)]: Done 138 tasks      | elapsed:   32.9s
[Parallel(n_jobs=-1)]: Done 341 tasks      | elapsed:  1.2min
[Parallel(n_jobs=-1)]: Done 624 tasks      | elapsed:  2.2min
[Parallel(n_jobs=-1)]: Done 989 tasks      | elapsed:  3.4min
[Parallel(n_jobs=-1)]: Done 1250 out of 1250 | elapsed:  4.2min finished
c:\users\jeevs\appdata\local\programs\python\python37\lib\site-packages\sklearn\model_selection\_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


RandomizedSearchCV(cv=5, error_score='raise-deprecating',
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    class_weight=None,
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_split=2,
                                                    min_weight_fraction_leaf=0.0,
                                                    n_estimators=100,
                                                    n_jobs=None,
 

In [16]:
rnd_search.best_params_

{'n_estimators': 138,
 'min_samples_split': 8,
 'min_samples_leaf': 5,
 'max_features': 'auto',
 'max_depth': 199,
 'bootstrap': False}

In [17]:
final_model = rnd_search.best_estimator_
forest_scores = cross_val_score(final_model, X_train, y_train, cv=5)
forest_scores.mean()

0.8328586875533507

## Preprocessing with new features

In [71]:
class DataSelector(BaseEstimator, TransformerMixin):
    def __init__(self, attribute_names):
        self.attribute_names = attribute_names
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        return X[self.attribute_names]
    
num_pipeline = Pipeline([
        ("select_numeric", DataSelector(["Age", "SibSp", "Parch", "Fare", "AgeRange", "FamilySize", "Title"])),
        ("imputer", SimpleImputer(strategy="median")),
    ])

#Imputer for string categories
class MostFrequentImputer(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.most_frequent_ = pd.Series([X[c].value_counts().index[0] for c in X],
                                        index=X.columns)
        return self
    def transform(self, X, y=None):
        return X.fillna(self.most_frequent_)

cat_pipeline = Pipeline([
        ("select_cat", DataSelector(["Pclass", "Sex", "Embarked"])),
        ("imputer", MostFrequentImputer()),
        ("cat_encoder", OneHotEncoder(sparse=False)),
    ])

preprocess_pipeline = FeatureUnion(transformer_list=[
        ("num_pipeline", num_pipeline),
        ("cat_pipeline", cat_pipeline),
    ])

X_train_copy = preprocess_pipeline.fit_transform(train_data_copy2)
y_train = train_data_copy["Survived"]

In [72]:
forest_clf = RandomForestClassifier(n_estimators=100, random_state=42)
forest_clf.fit(X_train_copy, y_train)
forest_scores = cross_val_score(forest_clf, X_train_copy, y_train, cv=5)
forest_scores.mean()

0.8080887315247571

In [73]:
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 2000, num = 50)]
max_features = ['auto', 'sqrt', 'log2']
max_depth = [int(x) for x in np.linspace(10, 220, num = 11)]
max_depth.append(None)
min_samples_split = [2, 3, 4, 5, 6, 7, 8, 9, 10]
min_samples_leaf = [1, 2, 3, 4, 5, 6, 7]
bootstrap = [True, False]

random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

rnd_search = RandomizedSearchCV(forest_clf, param_distributions=random_grid,
                                n_iter=300, cv=5, scoring='neg_mean_squared_error',
                                verbose=2, random_state=42, n_jobs=-1)
rnd_search.fit(X_train_copy, y_train)

Fitting 5 folds for each of 300 candidates, totalling 1500 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  17 tasks      | elapsed:    4.8s
[Parallel(n_jobs=-1)]: Done 138 tasks      | elapsed:   33.5s
[Parallel(n_jobs=-1)]: Done 341 tasks      | elapsed:  1.3min
[Parallel(n_jobs=-1)]: Done 624 tasks      | elapsed:  2.2min
[Parallel(n_jobs=-1)]: Done 989 tasks      | elapsed:  3.4min
[Parallel(n_jobs=-1)]: Done 1434 tasks      | elapsed:  5.1min
[Parallel(n_jobs=-1)]: Done 1500 out of 1500 | elapsed:  5.3min finished


RandomizedSearchCV(cv=5, error_score='raise-deprecating',
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    class_weight=None,
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_split=2,
                                                    min_weight_fraction_leaf=0.0,
                                                    n_estimators=100,
                                                    n_jobs=None,
 

In [74]:
rnd_search.best_params_

{'n_estimators': 1534,
 'min_samples_split': 10,
 'min_samples_leaf': 1,
 'max_features': 'log2',
 'max_depth': None,
 'bootstrap': True}

In [75]:
final_model = rnd_search.best_estimator_
forest_scores = cross_val_score(final_model, X_train_copy, y_train, cv=5)
forest_scores.mean()

0.8339319246453722

## Test dataset

In [78]:
#Add new features to test dataset
test_data_copy = test_data.copy()
test_data_copy = add_features(test_data_copy, True)

In [79]:
test_data_copy.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  FamilySize  AgeRange  \
0  34.5      0      0   330911   7.8292   NaN        Q           0         4   
1  47.0      1      0   363272   7.0000   NaN        S           1         5   
2  62.0      0      0   240276   9.6875   NaN        Q           0         5   
3  27.0      0      0   315154   8.6625   NaN        S           0         4   
4  22.0      1      1  3101298  12.2875   NaN        S           2         4   

  Title  
0     1  
1     2  
2     1  
3     1  
4     2

In [80]:
X_test = preprocess_pipeline.fit_transform(test_data_copy)

In [81]:
y_pred = final_model.predict(X_test)

In [82]:
submit_data = pd.DataFrame({
    'PassengerId' :test_data["PassengerId"],
    'Survived' :y_pred
})

submit_data.to_csv('titanic_submit7.csv', index = False)